### ISSUE TITLE
Is request.timeout supposed to override agent timeout?

### ISSUE THREAD

**Issue by doivosevic** [algesten/ureq#614]
I'm looking at https://github.com/algesten/ureq/blob/main/src/stream.rs#L342 and it seems as if the agent is first checked for timeout setting and only after is the unit checked. Wouldn't this go against https://docs.rs/ureq/latest/ureq/struct.Request.html#method.timeout where it says "Sets overall timeout for the request, overriding agent’s configuration if any."?

**Comment by algesten** [comment]
Hi @doivosevic! Thanks for looking at ureq!

It took me a second to figure this one out. You could argue the behavior is correct, but maybe a bit confusing and badly documented. 

The `unit.deadline` comes from here: https://github.com/algesten/ureq/blob/main/src/request.rs#L123 It prioritizes `self.timeout` over `agent.timeout`.

These called just `timeout` are _overall_ timeouts for the entire request/response.

However there's an override called `agent.connect_timeout` and that's the one you are looking at for https://github.com/algesten/ureq/blob/main/src/stream.rs#L342

So in summary. Request `timeout` correctly overrides Agent `timeout`, however for the connection step, the Agent `connect_timeout` might override them all.

**Comment by doivosevic** [comment]
I think last time ai looked, using ureq::post e.g. would just create default agent, and the default agent has connect_timeout set to 30 secs and so setting .timeout on a request cannot override connect_timeout and I have to create an agent first to set the connect_timeout. Did I get this right? I would expect I could use ureq::post and be able to control all timeouts, otherwise I feel like it loses a point using .timeout on Request

**Comment by algesten** [comment]
I see what you mean. The starting point is that the top level functions like`ureq::post` are there for convencience and that the the default agent have some reasonable defaults. For more detailed configuration, the user should set up their own agent.

That means your point raises some questions:

* Is 30 seconds connect_timeout a reasonable default?
* Is it right that `connect_timeout` overrides the overall `timeout`?
* And should it be possible to override `connect_timeout` on the request level?

These can all be debated.

**Comment by doivosevic** [comment]
I was confused because in docs it says "sets overall" so I expected that there wouldnt be any type of timeout which would be shorter than that after invoking that command. I had to dig into source code to figure out it wasnt my fault

**Comment by doivosevic** [comment]
Also, is https://github.com/algesten/ureq/blob/main/src/stream.rs#L132 error message "timed out reading response" correct? I can see this `time_until_deadline` being used in context where it isn't reading a response so is it really timing out while reading response in all cases? Sometimes maybe it timed out while filling a buffer or writing to stream?

**Comment by jsha** [comment]
Ah, this makes sense. For `Request::timeout`, where it says:

> Sets overall timeout for the request, overriding agent’s configuration if any.

That's a little confusing. So instead we should say something like:

> Sets overall timeout for the request, overriding configuration from [AgentBuilder::timeout](https://docs.rs/ureq/latest/ureq/struct.AgentBuilder.html#method.timeout). Note that, like AgentBuilder::timeout, this does not take precedence over [AgentBuilder::timeout_connect](https://docs.rs/ureq/latest/ureq/struct.AgentBuilder.html#method.timeout_connect).

Looking at the API more broadly it's perhaps a little confusing that `Request::timeout` exists but not `Request::timeout_connect`, `Request::timeout_read` and `Request::timeout_write`. I think, like @algesten said, we were trying to keep the API for Request simple and shunt more complex configurations to Agent. But if we find more people who find it confusing maybe we should revisit.

**Comment by algesten** [comment]
> we were trying to keep the API for Request simple and shunt more complex configurations to Agent. But if we find more people who find it confusing maybe we should revisit.

There's also an aspect of whether we should need to set something per request. I can't think of a case where `timeout_connect` would be something that needs to be individually adapted per request… maybe I'm overthinking though.

**Comment by jsha** [comment]
No, you're right; connect timeout is less likely to vary on a per-request
basis. It's starting to come back to me now; I think maybe the idea behind
making timeout (overall) available on request was that you might have some
requests you know will take a long time, for instance uploading a big file.
But you might want to make that request on the same connection you made
shorter requests on so you need to share the Agent.

**Comment by doivosevic** [comment]
> No, you're right; connect timeout is less likely to vary on a per-request basis. It's starting to come back to me now; I think maybe the idea behind making timeout (overall) available on request was that you might have some requests you know will take a long time, for instance uploading a big file. But you might want to make that request on the same connection you made shorter requests on so you need to share the Agent.

Yeah, I have some requests I make which take over 10 minutes and I don't have to reuse agent created somewhere, since it's not performance critical. I had no need for reusing a connection

**Comment by ex-nihil** [comment]
I created a PR yesterday related to the `agent.connect_timeout` and was also confused about the intended use of it.
#618

The problem I ran into was that a DNS first returned host was down but others worked.
But ureq always exited out on the first address (exiting on `time_until_deadline(deadline)?`) no matter what timeouts were configured.

**Comment by haverins** [comment]
I also found the timeout behavior confusing. Just to make sure, the total timeout for requests using this agent is not 25s but 15s (up to 10s to connect and rest to complete if connected):

```rust
let http_agent = ureq::AgentBuilder::new()
    .timeout_connect(std::time::Duration::from_secs(10))
    .timeout(std::time::Duration::from_secs(15))
    .build();
```

**Comment by jsha** [comment]
Yep, that's correct.

**Comment by amunra** [comment]
I had stumbled upon this bug independently and had filed a separate bug report: https://github.com/algesten/ureq/issues/758 (now marked as duplicate).

There I proposed a way forward:

---

# Suggested new behaviour

The `Request::timeout` should always be respected and override `AgentBuilder::timeout_connection`.
Moreover the precedence order here looks inverted: `AgentBuilder::timeout` should also override `AgentBuilder::timeout_connection` if set.

In other words, I suggest (pseudo-code):

```python
connection_timeout = min(
    agent.timeout_connection or 30s,
    agent.timeout or inf,
    request.timeout or inf)
```

The new `AgentBuilder::timeout_connection` docs would then read:
```
Timeout for the socket connection to be successful. If both this and `.timeout()` are both set, `.timeout()` takes precedence.

The default is 30 seconds.
```

This change would resolve the conflicting documentation and fix the feature so it becomes possible to set a per-request timeout.

---

How does this sound?

**Comment by algesten** [comment]
Closing since we're moving to ureq 3.x. The timeout values in ureq 3.x should be much more sensible: https://github.com/algesten/ureq/blob/0e3cbb912f0171b1aacd44ece22c704428800041/src/config.rs#L79-L137

**Comment by amunra** [comment]
A few questions:
  * What is the precendence of the new durations? For example, setting both `timeout_global` and `timeout_send_body`?
  * Can the timeouts be set on a per-request basis?

**Comment by algesten** [comment]
I guess I need to write more doc :)

`timeout_global` is the entire call, including every redirect, from start to finish.
`timeout_send_body` is the send body part.

Whichever has the least time left "wins.

Example:

`timeout_global: 2 sec`
`timeout_send_body: 1 sec`

If `resolve` and `connect` together take 1.5 sec, that leaves the lesser of global/send_body for sending the body, i.e. 0.5 sec.

> Can the timeouts be set on a per-request basis?

In some respects all timeouts are per-request. This is however only relevant for redirects (since that is the only time you have multiple request-response for the same call). So a timer like `timeout_resolve` is the timeout for every DNS resolve operation. If you get 3 redirects, you will encounter this timer 3 times. `timeout_global` would span the 3 requests overall, and `timeout_per_call` acts like global, but for an individual call.

**Comment by amunra** [comment]
Thank you for clarifying.

This issue was down to idiosyncrasies between the agent-level and request-level request timeout overrides. 

I'm looking at `request.rs` in the same 3.x [0e3cbb9](https://github.com/algesten/ureq/blob/0e3cbb912f0171b1aacd44ece22c704428800041/src/request.rs) commit.

It now no longer has the ability to set per-request timeouts -- unless I'm missing something.

We rely on being able to set the request timeout at this level since we calculate the timeout based on the amount of data we need to send.
For context, if uploading 1KiB this is expected to be quick, but if uploading 40MiB the timeout is adjusted accordingly.

We certainly don't want to have to create a new agent each time, since that would invalidate any active pooled TCP connections (which we want to maintain).

I'm not entirely sure how to achieve that with the new API.

**Comment by algesten** [comment]
Great feedback! Thanks!

Let's open an issue with that. I can do it unless you want to.
